site/content/posts/2023-08-09-v0.25.0-impersonation-proxy-with-external-certs.md

```diff
@@ -0,0 +1,337 @@
+    group: cert-manager.io
+---
+apiVersion: cert-manager.io/v1
+kind: ClusterIssuer
+metadata:
+  name: my-ca-issuer
+spec:
+  ca:
+    secretName: self-signed-ca-for-kind-testing
+---
+apiVersion: cert-manager.io/v1
+kind: Certificate
+metadata:
+  name: impersonation-serving-cert
+  namespace: pinniped-concierge
+spec:
+  secretName: impersonation-proxy-tls-serving-cert
+
+  duration: 2160h # 90d
+  renewBefore: 360h # 15d
+  subject:
+    organizations:
+    - Pinniped
+  isCA: false
+  privateKey:
+    algorithm: RSA
+    encoding: PKCS1
+    size: 2048
+  usages:
+  - server auth
+  dnsNames:
+  - impersonation-proxy-mtls.local
+  issuerRef:
+    name: my-ca-issuer
+    kind: ClusterIssuer
+    group: cert-manager.io
+
+EOF
+
+$ kubectl apply \
+    --filename self-signed-cert.yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+```
+
+Download the root (self-signed) CA's certificate so that it can be advertised as the CA bundle for the Concierge impersonation proxy:
+
+```shell
+$ kubectl get secret self-signed-ca-for-kind-testing \
+    --namespace cert-manager \
+    --output jsonpath="{.data.ca\.crt}" \
+    --kubeconfig kind-with-contour.kubeconfig.yaml \
+    | tee self-signed-ca-for-kind-testing.pem.b64
+    
+# Tip: Put the contents of self-signed-ca-for-kind-testing.pem.b64 into your copy buffer for a later step!
+```
+
+Now update the `CredentialIssuer` to use the impersonation proxy (which is disabled on kind by default):
+
+```shell
+$ kubectl edit credentialissuer pinniped-concierge-config \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+# Make sure that the spec has the following values:
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: # paste the contents of the file self-signed-ca-for-kind-testing.pem.b64
+        secretName: impersonation-proxy-tls-serving-cert
+...
+# Now save and close the text editor
+
+# Confirm that the CredentialIssuer looks as expected
+$ kubectl get credentialissuers pinniped-concierge-config \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+...
+  spec:
+    impersonationProxy:
+      externalEndpoint: impersonation-proxy-mtls.local
+      mode: enabled
+      service:
+        annotations:
+          # Ignore any annotations
+        type: ClusterIP
+      tls:
+        certificateAuthorityData: LS0tLUJFR0l..........
+        secretName: impersonation-proxy-tls-serving-cert
+...
+
+# Confirm that the ClusterIP service for the impersonation proxy was automatically created (may take a minute)
+$ kubectl get service pinniped-concierge-impersonation-proxy-cluster-ip \
+    --namespace pinniped-concierge \
+    --output yaml \
+    --kubeconfig kind-with-contour.kubeconfig.yaml
+
+# Configure a webhook authenticator to tell Concierge to validate static tokens using the installed local-user-authenticator
```

Comment:
Configure a webhook authenticator to tell Concierge to validate static tokens using the installed local-user-authenticator.  When we installed the Pinniped `local-user-authenticator`, we created a service called `local-user-authenticator` in the `local-user-authenticator` namespace.  We previously retrieved the `Secret` named `local-user-authenticator-tls-serving-certificate` so that we could use it to configure this `WebhookAuthenticator` to use that certificate.  Note that we did not generate this certificate via `cert-manager`, this is still a self-signed certificate created by Pinniped.

Reply:
The `endpoint` here is referenced via Kubernetes DNS in the format `<namespace>.<service-name>.svc` targeting the `/authenticate` endpoint of the `local-user-authenticator`.

Reply:
resolving locally